crates/plugin/src/stats.rs

```diff
@@ -0,0 +1,282 @@
+use std::collections::HashSet;
+use std::sync::{mpsc, Arc};
+
+use crate::prelude::*;
+use indexer_rabbitmq::geyser::{Message, SlotStatistics};
+use solana_program::instruction::CompiledInstruction;
+use solana_sdk::transaction::SanitizedTransaction;
+use solana_transaction_status::TransactionStatusMeta;
+
+use crate::sender::Sender;
+
+const STAT_REQ_BUFFER_SIZE: usize = 2500;
+const SLOT_BUFFER_SIZE: usize = 25;
+const LAMPORTS_PER_SIG: u64 = 5000;
+
+///the message sent to a producer to process
+#[derive(Debug)]
+pub(crate) struct StatsRequest {
+    pub slot: u64,
+    pub stx: SanitizedTransaction,
+    pub meta: TransactionStatusMeta,
+    pub is_vote: bool,
+    pub is_err: bool,
+}
+
+///the main stats struct that holds everything the
+///stats thread needs to do its job
+#[derive(Debug)]
+pub(crate) struct Stats {
+    ///a quick shortcut into the slot_stats buffer for most recent slot
+    ///this should allow a hot path of code execution
+    most_recent_slot_stats: (u64, usize),
+    ///the buffer of slot stats; slots stats are sent to rabbitmq after they
+    ///reach the size of the buffer. Thus, it's important to have the buffer large
+    ///enough to catch trailing transactions, but small enough to send before
+    ///the block is confirmed (so confirmooor can send it upon confirmation)
+    slot_stats: [SlotStatistics; SLOT_BUFFER_SIZE],
+    ///the rabbitmq sender
+    producer: Arc<Sender>,
+    ///the tokio async runtime to use for sending messages
+    rt: Arc<tokio::runtime::Runtime>,
+    ///used to identify token program ixs
+    token_programs: HashSet<Pubkey>,
+}
+
+impl Stats {
+    /// Create a new stats thread and return a handle to send stats requests to it
+    pub fn create_publisher(
+        producer: Arc<Sender>,
+        rt: Arc<tokio::runtime::Runtime>,
+    ) -> mpsc::SyncSender<StatsRequest> {
+        let (tx, rx) = mpsc::sync_channel::<StatsRequest>(STAT_REQ_BUFFER_SIZE);
+        //we use a dedicated worker thread, we don't play in the async dancing sandbox
+        //that the producer message sender uses
+        rt.clone().spawn_blocking(move || {
+            let mut stats = Stats {
+                most_recent_slot_stats: Default::default(),
+                slot_stats: std::iter::repeat::<SlotStatistics>(SlotStatistics::default())
+                    .take(SLOT_BUFFER_SIZE)
+                    .collect::<Vec<SlotStatistics>>()
+                    .try_into()
+                    .unwrap(),
+                producer,
+                rt,
+                token_programs: Self::get_token_programs(),
+            };
+            //the thread's endless loop of message processing
+            while let Ok(req) = rx.recv() {
```

Comment:
Could this ever not be OK? I'm thinking blips which would end the loop and would never be started again until validator restart?

Reply:
If it is `Err`, that means the `Sender` side was closed, which means validator died or is shutting down (or in future solana versions, the plugin is restarting).
If the `process` method panics, then the channel receiver would be closed so the sender side would err out (plugin.rs:437) and stop the validator.

Reply:
(if this loop exits, the Receiver is closed, which means the Sender will fail sending)